### PR TITLE
Do not apply content-type header if no request body

### LINKED ISSUE #6647: [JAVA] Bug: Feign client applies Content-Type header when no request body is present
##### Description

Requests with no body parameters (ie, no request body) should not include a Content-Type header. It's not necessary, and it could cause problems with parsing the request at the server side.

##### Swagger-codegen version

2.3.0 (90010da61)

##### Swagger declaration file content or url

```yaml
swagger: '2.0'
basePath: /
schemes:
  - http
paths:
  '/toggle':
    post:
      summary: Toggle the switch
      description: If switch is off, toggle to on, and vice versa.
      responses:
        200:
          description: Success
```

##### Command line used for generation
```
java -jar modules/swagger-codegen-cli/target/swagger-codegen-cli.jar generate \
    -i example.yaml \
    -l java \
    --library feign \
    -o $OUTPUT_DIR
```
##### Steps to reproduce

Generate a Java client using the Feign library using an API spec that includes a POST path with no body parameters. Use the client to make a request, and observe that the request has a "Content-Type" header.

##### Related issues/PRs

Issue https://github.com/swagger-api/swagger-codegen/issues/5613, fixed in https://github.com/swagger-api/swagger-codegen/pull/5616.

##### Suggest a fix/enhancement

Only add a Content-Type header is body parameters are specified: https://github.com/swagger-api/swagger-codegen/pull/6648

### PR DESCRIPTION
**Pull request by bmordue**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [x] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming langauge.

### Description of the PR

The intent is to avoid adding a Content-Type header on requests made by a generated Java-feign client if the request body is empty. This is to fix #6647 

* Do not set `contentType` for requests that have no body params specified.
* In the Java-Feign template, avoid adding an empty `@Headers` annotation for "Content-Type" if `contentType` is not set.

### CODE REVIEW

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractJavaCodegen.java [hunk 1]
```diff
@@ -941,13 +942,16 @@ public void preprocessSwagger(Swagger swagger) {
...
                     if (parameter instanceof FormParameter) {
                         hasFormParameters = true;
                     }
+                    if (parameter instanceof BodyParameter) {
+                        hasBodyParameters = true;
+                    }
                 }
-              //only add content-Type if its no a GET-Method
-                if(path.getGet() != null || ! operation.equals(path.getGet())){
+                if (hasBodyParameters){
```
Comment (wing328) [inline modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractJavaCodegen.java]: @bmordue shall we keep the logic `path.getGet() != null || ! operation.equals(path.getGet())` ?

For POST using HTTP form (not JSON/XML string), we also need the content-type right?

**Comment by wing328** [comment]
Please run `./bin/java-petstore-all.sh` to update Java Petstore client so that CIs can verify the result.

File: modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractJavaCodegen.java (hunk 1 above)
Comment (bmordue) [reply modules/swagger-codegen/src/main/java/io/swagger/codegen/languages/AbstractJavaCodegen.java]: Thanks, I've made sure the default content-type for form data gets applied too.

I don't think the original logic was correct: any path that includes a GET operation would have a default content-type applied, which doesn't seem right. I've added some tests that demonstrate this (ac799875265027d330e10a0efd485b9b0dfb3452).

**Comment by bmordue** [comment]
Apologies for my slow reply, I've been travelling. I ran the update script and committed the changes.

**Comment by wing328** [comment]
cc @bbdouglas @JFCote @sreeshas @jfiala @lukoyanov @cbornet

**Comment by JFCote** [comment]
@wing328 Is it normal that this generator use "gson" ? AFAIK, most if not all generator here use jackson? 
If that is ok, it looks good to me but I don't know Feign at all :)

**Comment by wing328** [comment]
cc @davidgri for review as well since the previous if-condition has been revised.

**Comment by wing328** [comment]
@JFCote Java Feign client uses jackson:
```
    <!-- JSON processing: jackson -->
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-core</artifactId>
      <version>${jackson-version}</version>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-annotations</artifactId>
      <version>${jackson-version}</version>
    </dependency>
    <dependency>
      <groupId>com.fasterxml.jackson.core</groupId>
      <artifactId>jackson-databind</artifactId>
      <version>${jackson-version}</version>
    </dependency>
    <dependency>
      <groupId>com.github.joschi.jackson</groupId>
      <artifactId>jackson-datatype-threetenbp</artifactId>
      <version>${jackson-threetenbp-version}</version>
    </dependency>
    <dependency>
      <groupId>org.apache.oltu.oauth2</groupId>
      <artifactId>org.apache.oltu.oauth2.client</artifactId>
      <version>${oltu-version}</version>
    </dependency>
```

File: samples/client/petstore/java/retrofit2rx2/src/main/java/io/swagger/client/JSON.java [hunk 2]
```diff
@@ -39,8 +46,42 @@
     private OffsetDateTimeTypeAdapter offsetDateTimeTypeAdapter = new OffsetDateTimeTypeAdapter();
     private LocalDateTypeAdapter localDateTypeAdapter = new LocalDateTypeAdapter();
 
+    public static GsonBuilder createGson() {
```
Comment (JFCote) [inline samples/client/petstore/java/retrofit2rx2/src/main/java/io/swagger/client/JSON.java]: @wing328 I'm talking about this. You can see use of Gson in other modifier files too.

Comment (cbornet) [reply samples/client/petstore/java/retrofit2rx2/src/main/java/io/swagger/client/JSON.java]: The retrofit generator does use gson. I guess this change is due to regenerating all the Java samples.

Comment (JFCote) [reply samples/client/petstore/java/retrofit2rx2/src/main/java/io/swagger/client/JSON.java]: ok no problem then. :)

**Comment by wing328** [comment]
If no further question or feedback on this, I'll merge it on coming Monday/Tuesday.

**Pull request merged**